commons/src/config/mod.rs

```diff
@@ -0,0 +1,75 @@
+use std::{path::PathBuf, collections::{HashMap, HashSet}};
+
+
+
+pub struct Config {
+  pub preamble: Preamble,
+  pub merging: MergingConfig,
+  pub input: PathBuf,
+}
+
+pub struct MergingConfig {
+  pub ignore_preamble: bool,
+}
+
+pub struct Preamble {
+  pub output: Output,
+  pub metadata: Metadata,
+  pub cite: Citedata,
+  pub render: RenderConfig,
+  pub i18n: I18n,
+}
+
+pub struct I18n {
+  pub lang: Option<String>,
+  pub langs: HashSet<String>,
+}
+
+
+pub struct RenderConfig {
+  /// K = element/attribute name, V = `true` to ignore
+  pub ignore: HashMap<String, bool>, 
```

Comment:
having „bool“ as value helps with negated ignores. since general types could be ignored, but specific types negated again. only having a hashset would make it hard to keep track of this. especially when merging preambles.

Reply:
What is the difference between ignored and negated?

Reply:
using `!` as first symbol in a line negates the ignore statement. meaning the type should not be ignored. this is useful to ignore more general types and only allow more specific ones. this is similar to .gitignore files

Reply:
How would that work? Ignore heading but allow heading with levels 1 and 2 as an example? 

Either way, `true/false` suggests that `ignore` has only on meaning: either ignore it or don't. We should try to improve this. For example we could use some other field name for this, like `render_options` or similar, and use an `enum` for value: 

```rust
enum RenderOption {
    Skip, // won't appear in the output
    Render, // will appear in the output
    // possibly others, like render but as original input, do not convert to html etc. 
}
```

That way we can clearly indicate what we mean. This may seem like an overkill, but remember this is clear to us (now), in the future it might not be so obvious for others or even us.

Reply:
As an example

```
heading
!heading-1
```

With this, all headings except of level 1 must be ignored.

I think `true/false` is reasonable here. `true` meaning `must be ignored` and `false` meaning `must not be ignored`.

**Note:** Only types set in ignore files are entered in the hashmap. If no entry of a type is found in the map, default rules apply (which is to **not** ignore). To decide wehter an element should be rendered or not, search for its type in the map. if it is not found, search for any parent type in the map. if none is found, render it.

I chose `ignore` as name, since the config also serves as CLI arguments. "Ignorefiles" are common for other tools, which should make it easier for users to understand the intended behavior.

Reply:
That is clear to me, however the internal representation is not that clear. 

`true` and `false` are not obvious. It's a little confusing when declared like that. I would either make the value an enum so we can explicitly mark what has to be done, OR when parsing ignorefile we insert every time explicitly, and remove negations. 

So for example

```
heading
```

would add `heading-1, heading-2, ..., heading-6` into the ignore `HashSet`. 

However
```
heading
!heading-2
```

would add `heading-1, ..., heading-6` but then for `!heading-2` remove the `heading-2`. That way we don't need any value for ignores, and when rendering we can simply always check what type is the block we're trying to render and proceed accordingly.

Reply:
My first attempt also was a hashset, but this leads to troubles with merging preambles.
Since every preamble might have their own ignore file, one could overwrite set ignores of "higher" levels.
e.g. ignoring all `attributes` per CLI, but in a preamble of an inserted file, one could negate this, resulting in attributes being allowed again.

Reply:
I just would not allow preamble to override that. For example, use case might be something like github rendering. If they want to ignore some stuff, the users should not be able to override that using preamble.

Reply:
That is why you may only **add** types to ignore, but negation is only possible at top level.
At least that is the plan for now.
Having the option to set ignores per preamble is kind of mandatory, if you do not want to always specify an ignore file per CLI

Reply:
Just to summarize for future readers. We discussed this internally and decided the following concept: 

* in one `ignorefile` users can ignore and negate ignores. this produces a `HashSet<T>` of ignored types. Exact representation will be decided later
* `HashSet<T>` contains **only ignored** types, no negations
* merging two instances of ignores stored as `HashSet<T>` is additive -> only new ignores can be added. Negations do not exist, so they cannot affect previous file
* consequently, ignores cannot be overridden, but negations can be overridden

i.e. a potential use case is some hosting platform (like Github) has ignores setup for every Unimarkup document. User uploaded Unimarkup documents **can** contain more ignores, but they cannot negate already ignored types by the hosting platform. The above described concept enables this use case.